### PR TITLE
AGM implementation of log(x) (Draft)

### PR DESCRIPTION
**Pull request by Dim131**
Draft implementation for the AGM for log(x).

### CODE REVIEW

**Comment by ckormanyos** [comment]
Great early start, Dimitris.

I have just setup my environment to examine this PR.

I am starting off with this code below. The answer is correct. I will report on the timing characteristics and review the code shortly.

Then we move on to some edge cases...

Kind regards, Chris

```
#include <chrono>
#include <limits>
#include <iomanip>
#include <iostream>

#include <boost/math/constants/constants.hpp>
#include <boost/multiprecision/cpp_bin_float.hpp>

namespace boost_gsoc2020
{
  template <class clock_type>
  struct stopwatch
  {
  public:
    typedef typename clock_type::duration duration_type;

    stopwatch() : m_start(clock_type::now()) { }

    stopwatch(const stopwatch& other) : m_start(other.m_start) { }

    stopwatch& operator=(const stopwatch& other)
    {
      m_start = other.m_start;
      return *this;
    }

    ~stopwatch() { }

    duration_type elapsed() const
    {
      return (clock_type::now() - m_start);
    }

    void reset()
    {
      m_start = clock_type::now();
    }

  private:
    typename clock_type::time_point m_start;
  };

  using big_float_type = boost::multiprecision::number<boost::multiprecision::cpp_bin_float<4000>,
                                                       boost::multiprecision::et_off>;

  using stopwatch_type = stopwatch<std::chrono::high_resolution_clock>;
}

int main()
{
  boost_gsoc2020::stopwatch_type my_stopwatch;

  boost_gsoc2020::big_float_type x = 10 + boost::math::constants::third<boost_gsoc2020::big_float_type>();

  // Compare with N[Log[31/3], 4000]
  my_stopwatch.reset();

  boost_gsoc2020::big_float_type y = log(x);

  const auto execution_time =
    std::chrono::duration_cast<std::chrono::duration<float>>(my_stopwatch.elapsed()).count();

  std::cout << std::setprecision(std::numeric_limits<boost_gsoc2020::big_float_type>::digits10)
            << y
            << std::endl;

  std::cout << "execution_time: " << std::setprecision(3) << execution_time << "s" << std::endl;
}
```

**Comment by ckormanyos** [comment]
> The answer is correct.

Actually, on the PR branch, I am getting the wrong numerical result.

I will investigate and see if my environment is wrong or something else is going on.

Kind regards, Chris.

**Comment by ckormanyos** [comment]
> I will investigate and see if my environment is wrong or something else is going on.

- There was a tiny problem determining the parameter `m`, which is now corrected using `std::numeric_limits<number<T> >`
- Also a test case with one single data point at 10k digits has been added.
- Performance improvement for log is noted with AGM above about 5000 decimal digits.

Good start, dimitris. Let's look a little bit more closely at the performance and see if we can optimize a bit.

Kind regards, Chris

File: include/boost/multiprecision/detail/functions/pow.hpp [hunk 1]
```diff
@@ -527,12 +523,14 @@ typename std::enable_if<should_use_log_agm<T>::value>::type eval_log(T& result_x
   // lexical_cast would do the trick.
```
Comment (Dim131) [inline include/boost/multiprecision/detail/functions/pow.hpp]: I used a slightly different approach based on the exponent of the epsilon threshold. Does this look reasonable to you?

File: include/boost/multiprecision/detail/functions/pow.hpp [hunk 2]
```diff
@@ -542,23 +540,17 @@ typename std::enable_if<should_use_log_agm<T>::value>::type eval_log(T& result_x
     // say 3 or 4.
```
Comment (Dim131) [inline include/boost/multiprecision/detail/functions/pow.hpp]: I switched to comparing the exponents of the difference a_k - b_k and the target tolerance.

File: include/boost/multiprecision/detail/functions/pow.hpp [hunk 3]
```diff
@@ -542,23 +540,17 @@ typename std::enable_if<should_use_log_agm<T>::value>::type eval_log(T& result_x
...
-    eval_divide(cp_ak, bk);
-    if (x.compare(0.0) < 0) {
-      cp_ak.negate();
-    }
-    T one(1.0);
-    eval_subtract(one, cp_ak);
-    if (one.compare(1.0) < 0) {
-      one.negate();
-    }
+    eval_subtract(cp_ak, bk);
```
Comment (Dim131) [inline include/boost/multiprecision/detail/functions/pow.hpp]: How could we avoid the subtraction a_k - b_k?

**Comment by ckormanyos** [comment]
> How could we avoid the subtraction a_k - b_k?

I was writing too fast. I suppose we can't avoid that. We do, in fact, need some kind of representation of the closeness.

Good catch.

**Comment by ckormanyos** [comment]
The Log AGM is starting to look really good.

A preliminary check with 10,000 decimal digit run result in:

```
Log AGM branch:
execution_time: 1.41s
closeness: 7.27e-9995

master branch:
execution_time: 6.4s
closeness: 1.14e-10000
```

**Comment by ckormanyos** [comment]
Nice work on the first few rounds of optimization, Dimitris.

This preliminary implementation is, from the viewpoint of performance, about where I expect it to be.

We see a nice and detectable speedup at 1,000 digits, probably meaning a speedup kicks in above 500 digits or so. In this sense, log calculations around 500 to 5,000 digits are expected to get a really nice performance improvement (and of course higher digits). But that 500 to, say, around 5,000 digits is the real _sweetspot_ for our new Karatsuba multiplication, whereas above a few thousand digits gets us into the future realm of Toom-Cook and FFT mul

```
Log AGM branch:
execution_time: 0.0146s
closeness: 1.11e-998

master branch:
execution_time: 0.0213s
closeness: 2.38e-1000
```

**Comment by ckormanyos** [comment]
Please see my off-site PM for more detailed comments.

**Comment by ckormanyos** [comment]
Nice work in these commits.
One comment of potential interest added.
Will address pending detailed questions via off-site mail.

**Comment by ckormanyos** [comment]
Great work on the tests and benchmarks.

The series of Benchmarks tells numerous stories.

- With the new Karatsuba mul, `cpp_bin_float` is, as it should, pulling way ahead of `cpp_dec_float`.
- Below 5 or 6 thousand decimal digits, `cpp_bin_float` is significantly slower than than GMP, but only ba a few factors, which makes sense because the _gc_ version (no benchmarks shown) of GMP itself is slower than the _core2_ version by a few factors. This makes sense because both `cpp_bin_float` and the _gc_ versoin of GMP use C++instead of assembly.
- Yet as the digit count climbs to 10,000, `cpp_bin_float` is, as it should, falling back to a factor of 10 behind GMP as the higher level multiplication Toom-Cook and maybe even FFT begin to kick in.
- Lightning fast is _MPFR_ showing inspirational performance on elementary transendental functions. Reverse engineering to see how in some areas might prove to be interesting.

```
Benchmark cpp_bin_float
1000 : 0.0209s
2000 : 0.216s
3000 : 0.653s
4000 : 1.21s
5000 : 2.12s
6000 : 3.56s
7000 : 4.77s
8000 : 6.66s
9000 : 9.07s
10000 : 12s
Benchmark cpp_dec_float
1000 : 0.0573s
2000 : 0.519s
3000 : 1.48s
4000 : 3.85s
5000 : 6.8s
6000 : 11.2s
7000 : 21.2s
8000 : 30s
9000 : 41s
10000 : 53.2s
Benchmark gmp
1000 : 0.0042s
2000 : 0.03s
3000 : 0.0807s
4000 : 0.157s
5000 : 0.277s
6000 : 0.43s
7000 : 0.652s
8000 : 0.882s
9000 : 1.16s
10000 : 1.41s
Benchmark mpfr
1000 : 0.000345s
2000 : 0.000889s
3000 : 0.00164s
4000 : 0.00238s
5000 : 0.00335s
6000 : 0.00424s
7000 : 0.00542s
8000 : 0.00648s
9000 : 0.00773s
10000 : 0.00909s
```

**Comment by Dim131** [comment]
Some benchmarking results for the Newton-Raphson implementation of exp

```
CPP dec 1'000
execution_time: 0.243s
closeness: 5.9e-1007
execution_time: 0.127s
closeness: 2.01e-1003
execution_time: 4e-07s
closeness: 0
execution_time: 0.131s
closeness: 2.01e-1003
execution_time: 0.126s
closeness: 3.55e-1002
CPP dec 5'000
execution_time: 5.86s
closeness: 3.96e-5003
execution_time: 3.23s
closeness: 3.98e-5004
execution_time: 4e-07s
closeness: 0
execution_time: 3.29s
closeness: 3.98e-5004
execution_time: 3.03s
closeness: 3.69e-5005
CPP bin 1'000
execution_time: 0.452s
closeness: 2.64e-990
execution_time: 0.276s
closeness: 4.86e-998
execution_time: 3e-07s
closeness: 0
execution_time: 0.281s
closeness: 4.87e-998
execution_time: 0.538s
closeness: 2.66e-998
CPP bin 5'000
execution_time: 13s
closeness: 1.78e-4997
execution_time: 7.34s
closeness: 2.18e-4998
execution_time: 5e-07s
closeness: 0
execution_time: 7.35s
closeness: 2.18e-4998
execution_time: 7.59s
closeness: 2.16e-4998
```

**Comment by ckormanyos** [comment]
Nice work!

**Comment by ckormanyos** [comment]
A good start on exp(x).
The answers are correct, but the timing performance is something we should take a closer look into.

**Comment by Dim131** [comment]
Timing results for the sqrt function (the first two results are most significant):
```
(Karatsuba square root)
execution_time: 0.00243s
closeness: 0

execution_time: 0.000897s
closeness: 0

Tested powers 2^i for i = 0, 2, ..., 256
execution_time: 6e-07s
closeness 0

execution_time: 5e-07s
closeness: 0

execution_time: 7.13e-05s
closeness: 0

execution_time: 3e-07s
closeness: 0

execution_time: 3e-07s
closeness: 0

execution_time: 4e-07s
closeness: 0

execution_time: 4e-07s
closeness: 0

execution_time: 4e-07s
closeness: 0

execution_time: 6e-07s
closeness: 0

execution_time: 6.5e-06s
closeness: 0

execution_time: 5.1e-06s
closeness: 0
```
and
```
execution_time: 0.0206s
closeness: 0

execution_time: 0.0197s
closeness: 0

Tested powers 2^i for i = 0, 2, ..., 256
execution_time: 2e-07s
closeness 0

execution_time: 4e-07s
closeness: 0

execution_time: 0.00346s
closeness: 0

execution_time: 1.7e-06s
closeness: 0

execution_time: 9e-07s
closeness: 0

execution_time: 2.3e-06s
closeness: 0

execution_time: 5.4e-06s
closeness: 0

execution_time: 1.98e-05s
closeness: 0

execution_time: 5.9e-06s
closeness: 0

execution_time: 2.42e-05s
closeness: 0

execution_time: 6.5e-06s
closeness: 0
```

**Comment by ckormanyos** [comment]
I have added a test project called "test_sqrt".
This project verifies the vastly improved performance of the sqrt kara implementation.
For an independent timing of square root and numerical verification, this test project also uses my separate implementation of [wide_integer](https://github.com/ckormanyos/wide-integer).

For large digit counts, however, the issues that we discussed in private e-mailes can also be observed.

Observations:

- Verify speedup of about factor of 10-20 compared with Boost 1.73.
- When switching digits2 from 8192 to 16384, test failures arise.
- Test failures also at 32678 digits2 and beyond.
- When test failures are observed, approximately 1/2 test cases fail (Could this be an issue with 1 even/odd bit in the base-case?).

We will continue the discussions in private e-mails.

**Pull request merged**